src/storage/src/hummock/utils.rs

```diff
@@ -203,83 +266,82 @@ impl MemoryLimiterInner {
         false
     }
 
-    async fn require_memory(&self, quota: u64) {
-        let current_quota = self.total_size.load(AtomicOrdering::Acquire);
-        if self.permit_quota(current_quota, quota)
-            && self
-                .total_size
-                .compare_exchange(
-                    current_quota,
-                    current_quota + quota,
-                    AtomicOrdering::SeqCst,
-                    AtomicOrdering::SeqCst,
-                )
-                .is_ok()
-        {
-            // fast path.
-            return;
+    fn require_memory(self: &Arc<Self>, quota: u64) -> MemoryRequest {
+        let mut waiters = self.controller.lock();
+        let first_req = waiters.is_empty();
+        if first_req {
+            // When this request is the first waiter but the previous `MemoryTracker` is just release a large quota, it may skip notifying this waiter because it has checked `inflight_barrier` and found it was zero. So we must set it one and retry `try_require_memory` again to avoid deadlock.
+            self.pending_request_count.store(1, AtomicOrdering::Release);
         }
-        loop {
-            let notified = self.notify.notified();
-            let current_quota = self.total_size.load(AtomicOrdering::Acquire);
-            if self.permit_quota(current_quota, quota) {
-                match self.total_size.compare_exchange(
-                    current_quota,
-                    current_quota + quota,
-                    AtomicOrdering::SeqCst,
-                    AtomicOrdering::SeqCst,
-                ) {
-                    Ok(_) => break,
-                    Err(old_quota) => {
-                        // The quota is enough but just changed by other threads. So just try to
-                        // update again without waiting notify.
-                        if self.permit_quota(old_quota, quota) {
-                            continue;
-                        }
-                    }
-                }
+        // We must require again with lock because some other MemoryTracker may drop just after this thread gets mutex but before it enters queue.
+        if self.try_require_memory(quota) {
+            if first_req {
+                self.pending_request_count.store(0, AtomicOrdering::Release);
             }
-            notified.await;
+            return MemoryRequest::Ready(MemoryTrackerImpl::new(self.clone(), quota));
         }
+        let (tx, rc) = channel();
+        waiters.push_back((tx, quota));
+        // This variable can only update within lock.
+        self.pending_request_count
```

Comment:
Store once atomic value is lightly comparing with acquire lock for waiter queue

Reply:
I don't get your point. Checking `first_req` does not need to acquire lock, because we still hold the lock, and we already know the value of `first_req`. Does it harm to add a check to avoid an unnecessary atomic storing?

Reply:
I means that storing once atomic value may only cost no more than 1/10 compare to calling sender

Reply:
But it still has cost, right? We are not making a trade off between A and B, but only deciding whether A is necessary or not.

Does it hurt to add a simple check to avoid an unnecessary atomic storing? Adding a check does not increase any code complexity, and can reduce a unnecessary cost. I don't see any downside on it.

Reply:
This line is not needed if:
- We make `pending_request_count` a atomic bool.
- We only set the atomic bool to true on first request, which we already do in this PR.
- We only set the atomic bool to false when the waiter queue becomes empty in `may_notify_waiters`.